### DISCUSSION TITLE
feature request: set configuration from environment variables

### DISCUSSION THREAD

**Discussion by gibson042** [avajs/ava#3164] [Q&A]
I have a need to invoke the same tests differently in the same CI setup; sometimes using the default reporter and sometimes using TAP. This precludes the possibility of using configuration files, and the particular setup involves automatic invocation in subdirectories in a way that makes varying arguments impractical.

But environment variables _are_ passed down fully, so it would be very convenient if the executable understood e.g. `AVA_TAP` as indicating that the `tap` setting should be on unless overridden by an argument. And fortunately, this is extremely convenient with the [yargs](https://www.npmjs.com/package/yargs) library already used by [cli.js](https://github.com/avajs/ava/blob/647d3e15305ae5fdfccf3009fdd1ef18a82af6a1/lib/cli.js)—basically as simple as [`.env('AVA')`](https://github.com/yargs/yargs/blob/HEAD/docs/api.md#envprefix) to support all arguments.

**Comment by novemberborn** [comment]
I'm reluctant to add configuration magic through environment variables. Configuration files (the JS ones) do have access to `process.env` though, could you use that to enable TAP conditionally?

**Comment by gibson042** [comment]
Nice idea. It works, but is a bit awkward because the subdirectories have package.json files with ava configuration so I need to explicitly specify `--config` and target a file that parses $projectDir/package.json to add overrides. But I understand (and in fact share) your apprehension around environment variables, and will understand if this case is not sufficiently compelling.

**Reply by novemberborn** [reply]
As long as those `package.json` files don't contain an `"ava"` key, AVA will walk up the directory tree looking for a JS file. Which receives the subdirectory if you export a factory function. Does that help?

**Reply by gibson042** [reply]
No, the package.json files _do_ contain an "ava" key with relevant package-level configuration.

**Reply by novemberborn** [reply]
You could replace that with `ava.config.js` files that import from a parent file or somesuch. Convoluted, but manageable.

**Reply by gibson042** [reply]
Well, I already have a workaround as mentioned above, which is basically the inverse of that to avoid the convolution and protect against accidental omissions in the ava config files—"_explicitly specify `--config` and target a file that parses $projectDir/package.json to add overrides_". But it would sure be much nicer to instead invoke the test run with `AVA_TAP=1 ava`.
